### PR TITLE
Working team pallet. Stakes and extra hiring flow extrinsics.

### PR DESCRIPTION
**Pull request by shamil-gadelshin**
Comments: 
- added stake support for the hiring flow extrinsics
- added multiple extrinsics:
   - increase_stake
   - decrease_stake
   - slash_stake
   - withdraw_application
   - cancel_opening
- unstaking period is not implemented yet

### CODE REVIEW

File: runtime-modules/working-team/src/lib.rs [hunk 1]
```diff
@@ -209,33 +255,50 @@ decl_module! {
...
+            staking_account_id: T::AccountId,
+            description: Vec<u8>,
+            stake: Option<BalanceOfCurrency<T>>,
         ) {
-            // Ensure origin which will server as the source account for staked funds is signed
+            // Ensure origin which will server as the source account for staked funds is signed.
             let source_account = ensure_signed(origin)?;
 
-            // Ensure the source_account is either the controller or root account of member with given id
+            // Ensure the source_account is either the controller or root account of member with given id.
```
Comment (bedeho) [inline runtime-modules/working-team/src/lib.rs]: 1. Why are we checking the root account?
2. Why are we coupled to the membership module? 

There should just be a simple `is_member` check on the runtime trait.

File: runtime-modules/working-team/src/types.rs [hunk 2]
```diff
@@ -101,14 +113,88 @@ pub struct Worker<AccountId, MemberId> {
...
+
+    /// Verifies that the stake could be decreased to a given amount.
+    fn ensure_can_decrease_stake(
+        lock_id: LockIdentifier,
+        account_id: &T::AccountId,
+        new_stake: BalanceOfCurrency<T>,
+    ) -> DispatchResult;
+
+    /// Decreases the stake for to a given amount.
+    fn decrease_stake(
```
Comment (bedeho) [inline runtime-modules/working-team/src/types.rs]: Why do decrease and increase return dispatchable results, they should not be able to fail - just like slash, as they just set a lock on the given account. It resulting in possible errors _MUTATION SAFE_ markers.

File: runtime-modules/working-team/src/types.rs [hunk 3]
```diff
@@ -101,14 +113,88 @@ pub struct Worker<AccountId, MemberId> {
...
+    /// Unstaking period for the stake. Zero means no unstaking period.
+    pub unstaking_period: BlockNumber,
+}
+
+/// Defines abstract staking handler to manage user stakes for different activities
+/// like adding a proposal. Implementation should use built-in LockableCurrency
+/// and LockIdentifier to lock balance consistently with pallet_staking.
+pub trait StakingHandler<T: system::Trait + GovernanceCurrency> {
+    /// Locks the specified balance on the account using specific lock identifier.
+    fn lock(lock_id: LockIdentifier, account_id: &T::AccountId, amount: BalanceOfCurrency<T>);
```
Comment (bedeho) [inline runtime-modules/working-team/src/types.rs]: Can we drop these lock_ids ? I know you wanted to post pone, but it really does not make sense to me to have them even now

File: runtime-modules/working-team/src/types.rs [hunk 4]
```diff
@@ -101,14 +113,88 @@ pub struct Worker<AccountId, MemberId> {
...
+
+    /// Removes the specified lock on the account.
+    fn unlock(lock_id: LockIdentifier, account_id: &T::AccountId);
+
+    /// Verifies that stake could be placed using given account.
+    fn ensure_can_make_stake(
```
Comment (bedeho) [inline runtime-modules/working-team/src/types.rs]: The semantics of this call is too vague, the name is overly broad. A consequence appears to be that no check is made that the staker is providing an account that is actually under their control. The signature of this interface could not check this even in principle, as it does not have the identity of the staker.

I suggest we make things crystal clear by having two distinct calls
- `is_member_staking_account(member_id, account_id) -> bool`
- `is_account_free_of_conflicting_staking(account_id) -> bool`

File: runtime-modules/working-team/src/types.rs (hunk 2 above)
Comment (bedeho) [reply runtime-modules/working-team/src/types.rs]: Let me add to that: why are there even checkers and mutators even here?

I don't see a need for either.

**Review by bedeho — CHANGES_REQUESTED** [review]
- I know this is not from this PR, but: Why does `JobOpening` have `created` field?

File: runtime-modules/working-team/src/types.rs (hunk 3 above)
Comment (shamil-gadelshin) [reply runtime-modules/working-team/src/types.rs]: Ok.

File: runtime-modules/working-team/src/types.rs (hunk 4 above)
Comment (shamil-gadelshin) [reply runtime-modules/working-team/src/types.rs]: Ok.

File: runtime-modules/working-team/src/lib.rs (hunk 1 above)
Comment (shamil-gadelshin) [reply runtime-modules/working-team/src/lib.rs]: Fixed.

File: runtime-modules/working-team/src/types.rs (hunk 2 above)
Comment (shamil-gadelshin) [reply runtime-modules/working-team/src/types.rs]: I changed the semantics.

**Comment by shamil-gadelshin** [comment]
Among other fixes I remove abstract `ensure_* `methods from the `StakingHandler` trait and introduced fine grained methods. Also, `decrease_stake()` is infallible now.

**Review by bedeho — APPROVED** [review]
LGTM.

**Pull request merged**